### PR TITLE
revamp TransportRequest handlers to support Writeable

### PR DESCRIPTION
**Pull request by talevy**
This PR begins the long journey to deprecating Streamable.

The idea here is to add additional method signatures that
support Writeable.Reader, so that the work to migrate objects TransportMessage to
implement Writeable and not Streamable.

One example conversion is done in this PR: SimulatePipelineRequest.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineRequest.java [hunk 1]
```diff
@@ -41,7 +42,7 @@
 
 import static org.elasticsearch.ingest.IngestDocument.MetaData;
 
-public class SimulatePipelineRequest extends ActionRequest {
+public class SimulatePipelineRequest extends ActionRequest implements Writeable {
 
     private String id;
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineRequest.java]: These can all become `final`?

File: core/src/main/java/org/elasticsearch/transport/TransportMessage.java [hunk 2]
```diff
@@ -22,14 +22,18 @@
...
 import org.elasticsearch.common.transport.TransportAddress;
 
 import java.io.IOException;
 
-public abstract class TransportMessage implements Streamable {
+public abstract class TransportMessage implements Streamable, Writeable {
 
     private TransportAddress remoteAddress;
 
+    public TransportMessage() {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/transport/TransportMessage.java]: a single, explicit empty ctor should not be necessary (that is what the default ctor is)

File: core/src/main/java/org/elasticsearch/transport/TransportRequest.java [hunk 3]
```diff
@@ -39,6 +39,11 @@
     public TransportRequest() {
     }
 
+    public TransportRequest(StreamInput in) throws IOException {
+        super();
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/transport/TransportRequest.java]: explicit call to default ctor is not necessary

File: core/src/test/java/org/elasticsearch/action/ingest/SimulatePipelineRequestTests.java [hunk 4]
```diff
@@ -89,4 +86,10 @@ public void testSerializationWithXContentBwc() throws IOException {
             }
         }
     }
+
+    public void testUnsupportedReadFrom() {
+        SimulatePipelineRequest request = new SimulatePipelineRequest();
+        Exception exception = expectThrows(UnsupportedOperationException.class,() -> request.readFrom(null));
+        assertThat(exception.getMessage(), equalTo("usage of Streamable is to be replaced by Writeable"));
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/action/ingest/SimulatePipelineRequestTests.java]: I don't think we should be adding a test for each one of these. It is not part of our API (that these are returning UOE).

**Review by rjernst — COMMENTED** [review]
I left some comments. Looks good in general.

File: core/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineRequest.java (hunk 1 above)
Comment (talevy) [inline core/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineRequest.java]: I think not, because of the client usage

File: core/src/test/java/org/elasticsearch/action/ingest/SimulatePipelineRequestTests.java (hunk 4 above)
Comment (talevy) [inline core/src/test/java/org/elasticsearch/action/ingest/SimulatePipelineRequestTests.java]: OK cool. I was on the fence about it. figured I'd discuss it after the work has been done :) will happily remove

File: core/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineRequest.java (hunk 1 above)
Comment (talevy) [inline core/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineRequest.java]: Even if the transport-client request builder pattern is to be deprecated, I think it may be best for that to be done in its own context in separate PRs

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineRequest.java]: Ah sorry, I could not see all the setters in the review without expanding.

**Comment by talevy** [comment]
thanks, updated with changes reflecting the review

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by nik9000** [comment]
Awesome! I'm sorry I missed this yesterday. I'm excited to see the start.

**Comment by dnhatn** [comment]
@talevy Is it possible to backport this to 6.x?

**Comment by talevy** [comment]
I suppose, I guess I didn't think about it since it was a work-in-progress transitional interface. If you think I should, I'd be happy to. I can imagine one would want to leverage the writeable interface in new actions in 7.0 that need to be backported to 6.x, so these aspects of the constructor could potentially cause merging issues? Is this what you are referring to?

**Comment by dnhatn** [comment]
@talevy. Yes, this actually happened to me. I was using a new API in V7 but had to use the old one for 6.x.

**Comment by ywelsch** [comment]
I've backported this PR (minus changes to `SimulatePipelineRequest`) to 6.x and 6.2 (39f62de and 9df8da4), as backporting other PRs turned out to be difficult without this.
